### ISSUE TITLE
"Snapshot" of writable layers

### ISSUE THREAD

**Issue by gorbak25** [containerd/overlaybd#223]
### What would you like to be added?

I would like for the `overlaybd-tcmu` binary to implement a TCMU reconfigure handler which:
1. Temporarily pauses IO on the block device
2. Seals the current writable layer
3. Creates a new empty writable layer
4. Reconfigures itself so the sealed layer is now an lower layer and the upper layer points to the new empty layer
5. Resumes IO
Sealing is quick so the IO pause shouldn't be noticeable for users of the block device.

### Why is this needed for Overlaybd?

OverlayBD will be the main storage back-end for https://github.com/hocus-dev/hocus(we're currently finishing the migration from raw sparse files to OverlayBD). We would have 2 usecases for reconfiguration:
1. Hocus spawns long lived vm's(workspaces) which we would like to periodically backup to an OCI registry without recreating the vm's from scratch.
2. When accessing remote repositories we spawn a dedicated VM for the job which creates a new layer with changes from the remote repository, if re-configuring tcmu was supported we could reuse the same block device between pulls of remote repositories.

### Are you willing to submit PRs to contribute to this feature?

- [ ] Yes, I am willing to implement it.

**Comment by gorbak25** [comment]
> incurs inconsistency between disk content and that cached in kernel
Would data still be inconsistent when first flushing the kernel writeback buffer by invoking `sync` on the guest/host and then reconfiguring the device?

**Comment by lihuiba** [comment]
Basically yes, as the kernel will not be aware that you have changed (reconfigured) the disk.

**Comment by lihuiba** [comment]
> whether `sync` would act as a barrier

No, `sync` is only a memory-to-disk synchronization, i.e. writing dirty cache pages back to disk; it does not take care of the opposite direction by loading new data from disk. (And the disk does not even have a way to tell what has been changed!)

Dropping caches with `sysctl -w vm.drop_caches=3` before reconfiguration does not seem to ensure consistency. I suspect that it does not fully drop all kinds of caches, but I didn't carry out further investigation.

**Comment by liulanzheng** [comment]
reconfiguration seems easy to implement because libtcmu already supports it. But it must be used with caution due cache consistency problems.
Perhaps this set of commands can solve the cache consistency issue:
```bash
sync
echo 3 > /proc/sys/vm/drop_caches
blockdev --flushbufs $DISK
hdparm -F $DISK
```
from https://stackoverflow.com/questions/9551838/how-to-purge-disk-i-o-caches-on-linux

**Comment by lihuiba** [comment]
Anyway, `umount` and re-`mount` again are necessart.

**Comment by gorbak25** [comment]
I don't see why purging the read cache is required when the only thing which changes is the upper writable layer so the data on disk remains the same. I see that problems will arise if one changes the disk image entirely but if only the upper layer changes then the read caches are still ok.

**Comment by lihuiba** [comment]
> I don't see why

Because you "pulls changes from the remote repository". This may incur inconsistency even if you don't "changes the disk image entirely".
